### PR TITLE
Setting form - start to sync preferences with Setting, using trait

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Adds a trait class & starts to align Settings form with preferences form

Before
----------------------------------------
The settings form & the preferences form do the same thing but evolved separately. Settings uses metadata, preferences doesn't.

After
----------------------------------------
This adds a trait that can be shared & the parts of both that use metadata can use the shared section

Technical Details
----------------------------------------
Over time we should remove both forms in favour of a new clean one based on the trait. I think the trait would be merged into that form.

Note that one issue TBC is how to define 'serialize' on the form. In the DAO we use a string but the generator converts it to a constant. I used just a constant here.

I can break this down into some smaller commits.

Comments
----------------------------------------
@mattwire I think part of https://github.com/civicrm/civicrm-core/pull/12714 is resolving the form interaction. I took a look on the first setting.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by colemanw** [comment]
@eileenmcnaughton this needs rebase now that the other PR merged.

**Comment by eileenmcnaughton** [comment]
@colemanw I rebased it - but I also created https://github.com/civicrm/civicrm-core/pull/12744 as a more reviewable chunk - which is a very replicable extraction

**Comment by mlutfy** [comment]
jenkins, test this please

**Comment by mlutfy** [comment]
jenkins, test this please

File: CRM/Admin/Form/Preferences.php [hunk 1]
```diff
@@ -251,6 +259,14 @@ public function postProcess() {
...
   public function postProcessCommon() {
+    try {
+      $this->saveMetadataDefinedSettings($this->_params);
+      $this->filterParamsSetByMetadata($this->_params);
+    }
+    catch (CiviCRM_API3_Exception $e) {
+      CRM_Core_Session::setStatus($e->getMessage(), ts('Save Failed'), 'error');
+    }
+
     foreach ($this->_varNames as $groupName => $groupValues) {
```
Comment (eileenmcnaughton) [inline CRM/Admin/Form/Preferences.php]: going forwards we basically migrate settings out of varNames into protected $_settings per below

File: settings/Core.setting.php [hunk 2]
```diff
@@ -42,17 +42,19 @@
...
       'optionGroupName' => 'contact_view_options',
     ),
     'default' => '123456789101113',
     'add' => '4.1',
-    'title' => 'Viewing Contacts',
+    'title' => ts('Viewing Contacts'),
     'is_domain' => '1',
     'is_contact' => 0,
-    'description' => NULL,
+    'description' => ts("Select the tabs that should be displayed when viewing a contact record. EXAMPLE: If your organization does not keep track of 'Relationships', then un-check this option to simplify the screen display. Tabs for Contributions, Pledges, Memberships, Events, Grants and Cases are also hidden if the corresponding component is not enabled. Go to Administer > System Settings > Enable Components to modify the components which are available for your site."),
```
Comment (eileenmcnaughton) [inline settings/Core.setting.php]: ideally we also use this for the tpl help text rather than duplicating - I haven't looked at that as yet

**Comment by colemanw** [comment]
@eileenmcnaughton what forms should I try when testing this?

**Comment by eileenmcnaughton** [comment]
@colemanw so far literally only one setting is converted  - the top one on display preferences

screenshot 2018-09-03 12 19 46

But ideally you would tinker with converting a few more to test the methodology a bit - @mattwire put up a PR with a lot more in it https://github.com/civicrm/civicrm-core/pull/12714/files#diff-cd9367d6c33a74944159c2f7623db4a7R69

An end goal would be to start using code like SettingsForm.tpl in Display.tpl - replacing the hard coded settings with metadata

**Comment by eileenmcnaughton** [comment]
Re the tpl - see this commit https://github.com/mattwire/civicrm-core/commit/73dad3696f2d9272c334cedb2db220ff222be207 - although I'm making the case for aligning settings & preferences forms as part of this change

**Comment by mattwire** [comment]
@eileenmcnaughton When I test this there seems to be something wrong with the checkbox save/load logic.  Pledges and mailings are always unchecked and if I uncheck grants cases gets unchecked as well.

**Comment by eileenmcnaughton** [comment]
@mattwire just figured it out - the setDefaults was the issue - can you recheck?

**Comment by mattwire** [comment]
@eileenmcnaughton Now working fine.  Can you fix the two style warnings reported by jenkins and then we can merge.

**Comment by eileenmcnaughton** [comment]
unrelated fails

**Pull request merged**